My Site Dashboard - Implement quick link ribbons

AjeshRPai opened the pull request:
Fixes #16005

This PR adds the changes for implementing the Quick Link Ribbon in Dashboard(Home) tab. Quick link Ribbon should behave the same way as the quick link actions on the Site menu tab. The navigation from the Quick Link Ribbon should behave in the same way as Quick Actions on the Site Menu Tab. 

| | |  
|---|---|
| ![Screenshot_20220405_002932](https://user-images.githubusercontent.com/17463767/161613270-d57201db-7437-4e48-ad6b-67ce950dc4b3.png)  |  ![Screenshot_20220405_002852](https://user-images.githubusercontent.com/17463767/161613282-f91d3fba-8595-4493-8115-89d083646d22.png) |  


The above UI - Quick Link Ribbon should have the same behaviour as Quick Link Action Card in the site menu, which is shown below. 

<img src="https://user-images.githubusercontent.com/17463767/161615111-4bfcc454-b0ea-424d-a0d5-4389edd9fbc2.png" width="370" height="680">
 

To test:
- Go to My Site -> Me -> App Settings -> Test Feature Configuration.
- Turn the MySiteDashboardPhase2FeatureConfig feature flag ON.
- Relaunch the app.
- Login with an account having wp com sites and go to the My Site tab.
- Notice that in the Home tab, Quick link ribbons are shown 
- Notice that the order of the Quick link ribbons are Pages, Posts, Media and Stats
- Clicking on each quick link ribbon item should take the user to the same destination as the one in quick link action card. 


### Known Issues 
####  [Quick start focus point on quick link ribbon](https://github.com/wordpress-mobile/WordPress-Android/issues/16227)

## Regression Notes
1. Potential unintended areas of impact
- Quick Link action card not shown 
- Quick Link Ribbon does not navigate the User to the correct destination 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and Unit Test

3. What automated tests I added (or what prevented me from doing so)
Manual tests

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Adds: Quick Link Ribbons list UI and Styles

AjeshRPai committed:
Adds: Quick Link Ribbons data class, Builder, Params and Test

AjeshRPai committed:
Adds: QuickLinkRibbonsViewHolder and Viewholder bind in Adapter

AjeshRPai committed:
Adds: Quick Link Builder to Viewmodel

Adds the logic of building quick link ribbon in Viewmodel and changes in tests according to that

AjeshRPai committed:
Adds: Missing diff util

AjeshRPai committed:
Updates: XML for changes

AjeshRPai committed:
Updates: Logic for showing quick link and quick ribbon

AjeshRPai committed:
Updates: Hard coded values with dimen values

AjeshRPai committed:
Updates: Ripple selector color

AjeshRPai committed:
Fixes: Formatting

AjeshRPai committed:
Removes: redundant id

AjeshRPai committed:
Updates: Style used in Quick link ribbon

AjeshRPai committed:
Fixes: Unresolved style

AjeshRPai committed:
Fixe: Names of the variables and functions in tests

AjeshRPai committed:
Fixes:Quick link ribbons builder logic in CardsBuilder

AjeshRPai committed:
ADDS: Missing tests in CardsBuilderTest for Quick Link Ribbons

AjeshRPai committed:
Adds: Missing tests for quick link ribbons view

AjeshRPai committed:
Fixes: Ktlint issues

AjeshRPai committed:
Updates: Card builder logic dependency on the correct feature config

AjeshRPai committed:
Fixes: Ktlint issues and removes duplicate test

AjeshRPai committed:
Fixes: Quick link ribbon horizontal scroll view paddings

AjeshRPai committed:
Refactor: The order of quick link ribbons

AjeshRPai committed:
Fix: Failing tests

AjeshRPai committed:
Fix: Failing tests

AjeshRPai commented:
@osullivanchris  Pinging you here since this change needs a design review. I am looking forward to your thoughts and suggestions on the design.

zwarm reviewed (changes requested):
Thanks @AjeshRPai - nicely done. The Quick start ribbon shows as expected on the Home tab. 

💛  I have a small request, can you remove the "s" from your property names please? We only have one ribbon and `QUICK_LINK_RIBBONS`, `QuickLinkRibbonsViewHolder`, `QuickLinkRibbons*Click` etc., make it sound like there is more than 1 ribbon.

❓  Do you intend to add fling support for the ribbon in a separate PR? You can take a look at the implementation for the Dynamic Cards in #16104 .

I added "tracking" requirements as a separate line item in the parent issue.

Thanks 👍

AjeshRPai committed:
Adds: Fling support for quick link ribbons

AjeshRPai committed:
Adds: Comment for suppressing the lint

AjeshRPai committed:
Updates: Naming of QuickLinkRibbons to QuickLinkRibbon in all classes

AjeshRPai commented:
@zwarm  

Thanks for the review and for adding the tracking requirements to the parent issue 👍🏼 .

>I have a small request, can you remove the "s" from your property names please? We only have one ribbon and QUICK_LINK_RIBBONS, QuickLinkRibbonsViewHolder, QuickLinkRibbons*Click etc., make it sound like there is more than 1 ribbon.

😅 Initially, I added the "s" to the QuickLinkRibbon, thinking that the single element is Ribbon and the entire view is Ribbons. My Mistake, I Have removed the "s" in the commit - 93c378d

>Do you intend to add fling support for the ribbon in a separate PR? You can take a look at the implementation for the Dynamic Cards in https://github.com/wordpress-mobile/WordPress-Android/pull/16104.

Thanks for pointing this out. I thought of adding the behaviour in a separate PR, but after looking at your changes in the [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16104), I felt that the changes are minor. So I added the fling support in this commit - 1fcd737.

❓ @zwarm The fling support on the quick link ribbon feels slightly different from the fling than in the dynamic cards. I am not sure. It might be just me. 😄 . I feel this is because of the combination of Horizontal Scroll View and Parent RecyclerView.  I had used the `Horizontal Scroll View` for Quick Link Ribbon instead of `RecyclerView` since the elements are static. I am wondering whether I should change it to RecyclerView. Let me know your thoughts. 🤔

osullivanchris commented:
> @osullivanchris Pinging you here since this change needs a design review. I am looking forward to your thoughts and suggestions on the design.

This is looking good, nice job @AjeshRPai 

My only question is the icon color. I'm not sure how we should define it. We use different colors for icons in the main tab bar, the app/navigation bar, the site menu list, and now here. What color have you used? 😄

AjeshRPai commented:
Hey @osullivanchris, I have used the same colour as we use for the Icon in the site menu list and also in the quick link action card icon. 😄 . Let me know if I need to change that. 🤔 . 


| Quick Link Ribbon icon colour  | Quick link action icon colour   | Site menu icon colour  |  
|---|---|---|
| ![Screenshot_20220405-161839](https://user-images.githubusercontent.com/17463767/161738889-6c8f7ee3-ec47-4c9c-889f-4ae72feee1bc.jpg)  | ![Screenshot_20220405-161856](https://user-images.githubusercontent.com/17463767/161738898-8324a5e6-a39a-4381-ad3f-b7889e32e35c.jpg)  |  ![Screenshot_20220405-161936](https://user-images.githubusercontent.com/17463767/161738902-40dcb97d-dd27-411a-a17f-62faaee001ce.jpg) |

osullivanchris commented:
> Hey @osullivanchris, I have used the same colour as we use for the Icon in the site menu list and also in the quick link action card icon. 😄 . Let me know if I need to change that. 🤔 .

Ok cool thanks for the demo! Let's keep that. Thanks!

zwarm commented:
@AjeshRPai - Thanks for the "s" changes. I'm getting that warm and fuzzy feeling now. :)

> The fling support on the quick link ribbon feels slightly different from the fling than in the dynamic cards.
 
I think the issue may be the 'scroll view' container itself.  If you switched your component from a scrollview to a horizontal recyclerview you'll get better behavior for the fling. Take a look at `ReaderInterestsCardViewHolder` which is a recycler view within a  view pager.  

Or you can also try shifting the gesture detector off of the `ScrollView` an on to the `constraint layout` and adjusting the Y_BUFFER lower? But I'm not certain this will handle all cases of flinging, espeically the edge of view issues correctly.

tiagomar reviewed (commented):
👋  @AjeshRPai, it looks good to me.

I've tested:
- [x] Quick Start
- [x] Light/Dark modes
- [x] Portrait/Landscape
- [x] TalkBack
- [x] English, German and Arabic (RTL)
- [x] Small and big fonts
- [x] Multitouch

AjeshRPai committed:
Updates: Quick Link ribbon view with RecyclerView

Adds: Quick link ribbon recycler view layout
Adds: Quick link ribbon row layout
Adds: Quick link ribbon item list data class
Adds: QuickLinkRibbonItemViewHolder and QuickLinkRibbonItemAdapter
Updates: Quick link ribbon data class with list
Updates: Quick link ribbon builder with the necessary changes
Updates: Tests with the changes

AjeshRPai commented:
@zwarm I have replaced the `HorizontalScrollView` with a Horizontal `RecyclerView` in the commit - 77abad9. The fling animation looks better now. 👍🏼.😄 

I should have done this in the first place. 😄 . Thanks for the suggestion. Please have a look at the changes.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quicklinkribbons/QuickLinkRibbonViewHolder.kt:
```diff
@@ -4,36 +4,52 @@ import android.annotation.SuppressLint
+            if (adapter == null) {
+                layoutManager = LinearLayoutManager(context, RecyclerView.HORIZONTAL, false)
+                val readerInterestAdapter = QuickLinkRibbonItemAdapter()
+                adapter = readerInterestAdapter
```
@AjeshRPai - Probably want to rename this something other than `readerInterestAdapter` 😉

AjeshRPai replied:
@zwarm Good catch 😄 . I have removed the variable and assigned the adapter to the recycler view. Changes in 6ccec7c

zwarm reviewed (approved):
👋  Works like a charm; including the fling! 🙏  for all the updates. I have one change request re: a copy/paste name - see the individual comment. After you change that feel free to merge.

AjeshRPai committed:
Updates: the adapter initialization

AjeshRPai commented:
@zwarm Thanks for the review and suggestions. 👍🏼 🙌🏼 . I will merge this PR once the CI checks are completed successfully.
